[QA-894] Change large collections parameters and add validations

Mark-Gurevich opened the pull request:
This PR changes some parameters for testing large collections and adds verification methods for them. Also some filters for warnings were added.


## PR pre-checks (self review)
<!--- PR should be created as Draft, when CI finished and relevant checkboxes selected, add reviewers and then click on "Ready for review" button.-->
<!--- Put an `x` in all the boxes that apply or create PR and then click on all relevant checkboxes: -->
- [x] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [x] I didn't leave commented-out/debugging code
- [ ] I added the relevant `backport` labels
- [x] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [ ] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [x] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

Mark-Gurevich commented:
Warning for writing large collection.
```
2023-07-09 19:30:05.459 <2023-07-09 19:30:05.000>: (DatabaseLogEvent Severity.WARNING) period_type=one-time event_id=f892929f-ffd2-4b26-b3b9-38331a8c5cd1: type=WARNING regex=(^WARNING|!\s*?WARNING).*\[shard.*\] line_number=492888 node=longevity-large-collections-1h-qa-8-db-node-ffd6f51c-4
2023-07-09T19:30:05+00:00 longevity-large-collections-1h-qa-8-db-node-ffd6f51c-4  !WARNING | scylla[5786]:  [shard 9] large_data - Writing large collection large_collection_test/table_with_large_collection: .L/b/nums_list (292770 bytes) to me-709-big-Data.db
```

Mark-Gurevich commented:
Added ignore_large_collection_warning context manager.
Trying to fix the error in validation.

Mark-Gurevich commented:
1. Adjust the load
2. Check warning filter method
3. Change sct runner instance type

fgelcer commented:
> This requires some extra work.
> 
> 1. Fix the validation method for searching the system log for "writing large collection" message on the nodes.
> 2. Build logic to ignore these warning messages so events in sct would not be created

and, any progress with these?

Mark-Gurevich commented:
No progress

Mark-Gurevich commented:
https://jenkins.scylladb.com/job/scylla-staging/job/mark/job/longevity-large-collections-1h-test/

Mark-Gurevich commented:
As we encountered an issue with stop_events_analyzer timeout we decided to filter 2 kind of warnings:
1. "Writing large collection" - the warning from large_data when it writes a large collection.
2. `mutation_partition - Memory usage of unpaged query exceeds soft limit` - also happens when we have many large collections.
In the last run number 15 I see another ~200+ warnings:
```
2023-08-15 10:51:16.941 <2023-08-15 10:51:16.000>: (DatabaseLogEvent Severity.WARNING) period_type=one-time event_id=466b39ab-3a96-4f72-bab3-10f20699c52e: type=WARNING regex=(^WARNING|!\s*?WARNING).*\[shard.*\] line_number=160969 node=longevity-large-collections-1h-qa-8-db-node-40feab91-2
2023-08-15T10:51:16+00:00 longevity-large-collections-1h-qa-8-db-node-40feab91-2  !WARNING | scylla[5582]:  [shard 8] querier - Read 10 live rows and 1093 tombstones for system.large_cells partition key "large_collection_test:table_with_large_collection" {{1818737829577356380, pk{00156c617267655f636f6c6c656374696f6e5f74657374001b7461626c655f776974685f6c617267655f636f6c6c656374696f6e}}} (see tombstone_warn_threshold)
```

Anyway still the test "fails" with timeout on the stop_events_analyzer() method.
@fgelcer 
Can you please write the additional changes required that you mentioned?

Mark-Gurevich commented:
https://jenkins.scylladb.com/job/scylla-staging/job/mark/job/longevity-large-collections-threshold-12h-test/18/

roydahan commented on data_dir/large_collections.yaml:
```diff
@@ -55,17 +55,18 @@ columnspec:
 # 1448.1546878700494
 # >>> math.sqrt(1 * 1024 * 1024)
 # 1024.0
+# gaussian(min..max,mean,stdev)
```
@Mark-Gurevich are all the comments above still relevant?
If not, please add your comments explaining the size and population you used.

Mark-Gurevich replied:
https://github.com/scylladb/qa-tasks/issues/894#issuecomment-1713449685

`if gaussian, min = 6k, max = 12k, mid = 9k, diff = 13%`
This calculation gives us that most of the rows do not exceed the threshold, but only some do as you requested.
Anyway we still filter the warnings mentioned above.

roydahan replied:
Your answer doesn't answer what I asked.
I see some calculations above in the comments, but the numbers you're using are different from them.
So, I asked if these are still relevant.

And regarding to what you've answered about "size: gaussian(6000..12000,9000,13)", it would be good to add it in the comment, so the next person read it, will understand why you chose these numbers.

roydahan commented on jenkins-pipelines/longevity-large-collections-1h.jenkinsfile:
Please remove the 1h from the name.

Mark-Gurevich replied:
Removed the file.

roydahan commented on test-cases/longevity/longevity-large-collections-1h.yaml:
```diff
@@ -0,0 +1,25 @@
+test_duration: 60
+stress_cmd: [
+  "JVM_OPTION='-Xms8033M -Xmx8033M -Xmn400M' cassandra-stress user profile=/tmp/large_collections.yaml ops'(insert=3,read1=1,update1=1)' cl=QUORUM duration=20m -mode cql3 native -rate threads=8"
```
We have now "custom duration".
Set the duration by default to 12h.

Mark-Gurevich replied:
If we don't need the 1h test as we have custom duration I should totally remove this file.
We have another (old)file with 12h by default.
This file is also in this commit with added lines.
```
validate_large_collections: true
root_disk_size_runner: 150
```

roydahan commented on test-cases/longevity/longevity-large-collections-1h.yaml:
```diff
@@ -0,0 +1,25 @@
+test_duration: 60
```
Adjust to 13h by default.

Mark-Gurevich replied:
Removed the file

roydahan commented on test-cases/longevity/longevity-large-collections-1h.yaml:
```diff
@@ -0,0 +1,25 @@
+test_duration: 60
+stress_cmd: [
+  "JVM_OPTION='-Xms8033M -Xmx8033M -Xmn400M' cassandra-stress user profile=/tmp/large_collections.yaml ops'(insert=3,read1=1,update1=1)' cl=QUORUM duration=20m -mode cql3 native -rate threads=8"
```
Why is the "rate threads" so low?

Mark-Gurevich replied:
Removed this file.

roydahan commented on test-cases/longevity/longevity-large-collections-1h.yaml:
```diff
@@ -0,0 +1,25 @@
+
+instance_type_db: 'i3en.3xlarge'
+
+nemesis_class_name: 'NoOpMonkey'
```
Change to SisyphusMonkey

Mark-Gurevich replied:
Removed this file.

roydahan commented on test-cases/longevity/longevity-large-collections-1h.yaml:
```diff
@@ -0,0 +1,25 @@
+nemesis_seed: '012'
+nemesis_interval: 15
+
+user_prefix: 'longevity-large-collections-1h'
```
remove the 1h

Mark-Gurevich replied:
Removed this file.

roydahan commented on test-cases/longevity/longevity-large-collections-1h.yaml:
```diff
@@ -0,0 +1,25 @@
+nemesis_interval: 15
+
+user_prefix: 'longevity-large-collections-1h'
+space_node_threshold: 64424
```
remove

Mark-Gurevich replied:
Removed this file.

roydahan commented on test-cases/longevity/longevity-large-collections-1h.yaml:
```diff
@@ -0,0 +1,25 @@
+hinted_handoff: 'disabled'
+
+append_scylla_yaml: |
+  compaction_collection_items_count_warning_threshold: 9000
```
please add a comment explain what it does and why we need it.

Mark-Gurevich replied:
Removed this file.

roydahan commented on longevity_test.py:
```diff
@@ -78,6 +95,21 @@ def run_pre_create_keyspace(self):
+                    " AND table_name='table_with_large_collection' ALLOW FILTERING"
+            statement = SimpleStatement(query, fetch_size=10)
+            data = list(session.execute(statement))
+            assert len(data) >= 1, "Did not find expected row in system.large_cells"
```
Not sure if "assert" is the correct behaviour here, maybe "Error" event is better.

Mark-Gurevich replied:
Sorry for stupid question, but do you mean just to use `if` statement and print LOGGER.ERROR message ?
This is the main idea behind this test to check the table has an entry for large collections and relevant message appear in the logs.

roydahan replied:
This is not the main idea being this test, it's another validation you added to this test scenario which tests how scylla behaves with large collections.

So, error event is better IMO.

Mark-Gurevich replied:
Ok, then I change this to `if` and `self.log.error`.

fgelcer replied:
`self.log.error` is not the same as an `error event` FYI

Mark-Gurevich replied:
Changed to InfoEvent("message", severity=Severity.ERROR)

roydahan commented on longevity_test.py:
```diff
@@ -78,6 +95,21 @@ def run_pre_create_keyspace(self):
+        self.log.info("Verifying warning for large collections in logs on node: {}".format(node))
+        msg = "Writing large collection"
+        res = list(node.follow_system_log(patterns=[msg], start_from_beginning=True))
+        assert res, f"Did not find expected log message warning: {msg}"
```
same.
(Although I'm not sure it's really needed to verify also in logs)

Mark-Gurevich replied:
This was your request for this test as we have only 2 places where we can check it. The `system.large_cells` table and relevant message in logs.

roydahan replied:
We can keep it as long as it doesn't cause a performance issue for SCT to monitor these messages.

roydahan reviewed (changes requested):
See comments above.

Mark-Gurevich commented on test-cases/longevity/longevity-large-collections-1h.yaml:
```diff
@@ -0,0 +1,25 @@
+n_loaders: 1
+n_monitor_nodes: 1
+
+instance_type_db: 'i3en.3xlarge'
```
Removed this file.

Mark-Gurevich commented on test-cases/longevity/longevity-large-collections-1h.yaml:
```diff
@@ -0,0 +1,25 @@
+
+# Disable hinted handoff to avoid a known issue:
+# https://github.com/scylladb/scylla/issues/4935
+hinted_handoff: 'disabled'
```
Removed this file.

roydahan commented:
Please run the 12h test and attach its results here.

Mark-Gurevich commented:
> Please run the 12h test and attach its results here.

@roydahan 
I've already ran it.
https://jenkins.scylladb.com/job/scylla-staging/job/mark/job/longevity-large-collections-threshold-12h-test/18/
The test passed successfully.

soyacz reviewed (commented):
Please improve commit message (more overview and specify why and what instead of how) and PR message

Mark-Gurevich commented:
> Please improve commit message (more overview and specify why and what instead of how) and PR message

I've changed the title and description for the commit. Added overview as requested.
I saw some engineers in "Dev" write details in the commit message so followed their lead.

fruch commented on longevity_test.py:
```diff
@@ -78,6 +96,23 @@ def run_pre_create_keyspace(self):
+    def _run_validate_large_collections_in_system(self, node):
+        self.log.info("Verifying large collections in system tables on node: {}".format(node))
+        with self.db_cluster.cql_connection_exclusive(node=node) as session:
+            query = "SELECT * from system.large_cells WHERE keyspace_name='large_collection_test'" \
```
nitpick, I would have not filter for a specific table, so this validation can be use on other test case if needed
also we don't expect other tables to have large cells, do we ?

fruch reviewed (approved):
LGTM

soyacz commented on longevity_test.py:
```diff
@@ -35,6 +41,18 @@ class LongevityTest(ClusterTester, loader_utils.LoaderUtilsMixin):
     Test a Scylla cluster stability over a time period.
     """
 
+    def __init__(self, *args, **kwargs):  # pylint: disable=too-many-statements,too-many-locals,too-many-branches
+        super().__init__(*args)
+
+        # This ignores large_data warning messages "Writing large collection" for large collections to prevent
+        # creating SCT Events from these warnings.
+        # During large collections test thousands of warnings are being created.
+        self.validate_large_collections = self.params.get('validate_large_collections')
+        if self.validate_large_collections:
+            self.stack = contextlib.ExitStack()
+            self.stack.enter_context(ignore_large_collection_warning())
+            self.stack.enter_context(ignore_max_memory_for_unlimited_query_soft_limit())
```
I wonder if we need this play with context - looking at the `BaseFilter` entering context means just `self.publish()` and in case we won't need cancel filtering we could just create filter and publish like this:
```
DbEventsFilter(
            db_event=DatabaseLogEvent.WARNING,
            line="Writing large collection"
        )).publish()
```
I think it should do the job and reduce LoC/complication

We could do it in test itself, so we enable it as a part of the test flow

fruch replied:
this style was mostly for purpose of reusing context manager we already have and used in other places.
also it can stopped if we want at later point.

I think it's align nicely the way we use those filters, and not having to keep both option in mind.

soyacz replied:
if it's set like it is now, we can't stop at later point - playing with 'self.stack` to exit some context is quite complicated.

I'm just saying, this case context manager is more complicating than helping.

Mark-Gurevich commented:
Can we merge this ?

roydahan commented:
@soyacz can you please finalize review and if approved, please merge.

soyacz reviewed (approved):
LGTM

soyacz commented:
need one approval more to enable merge

fgelcer reviewed (approved):
LGTM
